hazelcast/src/main/java/com/hazelcast/query/impl/AbstractIndex.java

```diff
@@ -57,17 +58,24 @@
     private final boolean ordered;
     private final PerIndexStats stats;
 
+    /**
+     * Reference to the store if it is bound to the same partition as the index (local index), {@code null} otherwise.
+     */
+    private final StoreAdapter partitionStoreAdapter;
+
     private volatile TypeConverter converter;
 
     @SuppressFBWarnings("EI_EXPOSE_REP2")
     public AbstractIndex(String name, String[] components, boolean ordered, InternalSerializationService ss,
-                         Extractors extractors, IndexCopyBehavior copyBehavior, PerIndexStats stats) {
+                         Extractors extractors, IndexCopyBehavior copyBehavior, PerIndexStats stats,
+                         StoreAdapter partitionStoreAdapter) {
         this.name = name;
         this.components = components;
         this.ordered = ordered;
         this.ss = ss;
         this.extractors = extractors;
         this.copyBehavior = copyBehavior;
+        this.partitionStoreAdapter = partitionStoreAdapter;
```

Comment:
AFAIU, this store adapter reference is required for HD only, probably it's better to move it to HD-specific index implementation, not to save some memory, just to avoid wondering why this store adapter is always null.

BTW, the same applies to index reference passed to index stores, looks like it can be replaced with the store reference passed to HD index stores only.